File: 백준/Silver/1697. 숨바꼭질/숨바꼭질.cc
#include <bits/stdc++.h>
using namespace std;

int board[100005];
int dist[100005];

int main(void) {
	int n, k;
	cin >> n >> k;

	queue<int> Q;
	fill(dist, dist + 100005, -1);

	dist[n] = 0;
	Q.push(n);

	while (dist[k] == -1) {
		int cur = Q.front(); Q.pop();
		for (int nxt : {cur - 1, cur + 1, cur * 2}) {
			if (nxt < 0 || nxt > 100000) continue;
			if (dist[nxt] >= 0 ) continue;

			dist[nxt] = dist[cur] + 1;
			Q.push(nxt);
		}
	}

	cout << dist[k];
}

File: 백준/Silver/1697. 숨바꼭질/숨바꼭질.java


import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.ArrayDeque;
import java.util.Arrays;
import java.util.Queue;
import java.util.StringTokenizer;

public class Main {
	
	static int N, K;
	static int[] dist = new int[200005];
	static Queue<Integer> q = new ArrayDeque<>();
	
	public static void main(String[] args) throws Exception {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		
		StringTokenizer st = new StringTokenizer(br.readLine());
		
		N = Integer.parseInt(st.nextToken());
		K = Integer.parseInt(st.nextToken());
		
		Arrays.fill(dist, -1);
		
		q.offer(N);
		dist[N] = 0;
		
		while(!q.isEmpty()) {
			int cur = q.poll();

			if(cur == K) {
				System.out.println(dist[cur]);
				break;
			}
			
			
			for(int nxt : new int[] { cur + 1, cur - 1, cur * 2 } ) {
				if(nxt < 0 || nxt > 200000) continue;
				if(dist[nxt] >= 0) continue;		
				
				dist[nxt] = dist[cur] + 1;
				q.offer(nxt);
			}
			
		}
	}

}
